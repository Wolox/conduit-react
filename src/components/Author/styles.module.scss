@import 'scss/variables/_colors';

.container {
  position: relative;

  .container-author {
    align-items: center;
    display: flex;
    flex-direction: row;

    .image {
      border-radius: 30px;
      display: inline-block;
      height: 32px;
      margin-right: 5px;
      vertical-align: middle;
      width: 32px;
    }

    .username {
      color: $green-fern;
      display: block;
      font-size: 16px;
      font-weight: 600;
    }
  }

  .container-button {
    align-items: center;
    display: flex;
    height: 100%;
    justify-content: center;
    position: absolute;
    right: 15px;
    top: 0;

    .like {
      border-radius: 3px;
      border: 1px solid $green-fern;
      color: $green-fern;
      padding: 4px 6px;

      &:hover {
        background: $green-fern;
        color: $white;
      }
    }

    .active {
      background: $green-fern;
      color: $white;
    }

    .icon {
      margin-right: 3px;
    }
  }
}
