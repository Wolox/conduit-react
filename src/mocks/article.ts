import { ArticleResponse } from 'types/Article';

const date = new Date('2021-08-26T18:09:50.028Z');

export const MOCKED_ARTICLE_RESPONSE: ArticleResponse = {
  article: {
    title: 'Luigi Bros',
    slug: 'luigi-bros',
    body: '<h1>Luigi rocks</h1><p><strong>Mario and Bowser suck</strong></p>',
    createdAt: date,
    updatedAt: date,
    description: 'Luigi is better than Mario',
    tagList: [],
    author: {
      username: 'luigi_da_boss',
      bio: null,
      image: 'luigi-pic.png',
      following: true
    },
    favorited: false,
    favoritesCount: 77
  }
};
