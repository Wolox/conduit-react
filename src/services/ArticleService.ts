import { ApiResponse } from 'apisauce';

import api from 'config/api';
import type {
  ArticleResponse,
  Articles,
  ArticlesByAuthor,
  ArticlesFavorites,
  FavoritesAddRemove,
  NewPostPayload,
  Paginated,
  PaginatedByTag,
  CommentsResponse,
  CommentResponse,
  CommentPayload
} from 'types/Article';

const MAIN_PATH = '/articles';

export const articles = (payload: PaginatedByTag): Promise<ApiResponse<Articles>> =>
  api.get(MAIN_PATH, payload);

export const feed = (payload: Paginated): Promise<ApiResponse<Articles>> =>
  api.get(`${MAIN_PATH}/feed`, payload);

export const articlesByAuthor = (payload: ArticlesByAuthor): Promise<ApiResponse<Articles>> =>
  api.get(MAIN_PATH, payload);

export const articlesFavorites = (payload: ArticlesFavorites): Promise<ApiResponse<Articles>> =>
  api.get(MAIN_PATH, payload);

export const addRemoveFavorites = ({
  slug,
  isFavorite
}: FavoritesAddRemove): Promise<ApiResponse<ArticleResponse>> =>
  isFavorite ? api.delete(`${MAIN_PATH}/${slug}/favorite`) : api.post(`${MAIN_PATH}/${slug}/favorite`);

export const addNewPost = (payload: NewPostPayload): Promise<ApiResponse<ArticleResponse>> =>
  api.post(MAIN_PATH, payload);

export const updatePost = (params: {
  slug: string;
  payload: NewPostPayload;
}): Promise<ApiResponse<ArticleResponse>> => api.put(`${MAIN_PATH}/${params.slug}`, params.payload);

export const articleBySlug = (slug: string): Promise<ApiResponse<ArticleResponse>> =>
  api.get(`${MAIN_PATH}/${slug}`);

export const deleteArticleBySlug = (slug: string): Promise<ApiResponse<ArticleResponse>> =>
  api.delete(`${MAIN_PATH}/${slug}`);

export const commentsBySlug = (slug: string): Promise<ApiResponse<CommentsResponse>> =>
  api.get(`${MAIN_PATH}/${slug}/comments`);

export const addNewComment = (payload: CommentPayload): Promise<ApiResponse<CommentResponse>> =>
  api.post(`${MAIN_PATH}/${payload.slug}/comments`, payload.comment);

export const deleteComment = (params: { slug: string; id: number }): Promise<ApiResponse<void>> =>
  api.delete(`${MAIN_PATH}/${params.slug}/comments/${params.id}`);
