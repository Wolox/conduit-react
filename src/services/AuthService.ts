import { ApiResponse } from 'apisauce';

import api from 'config/api';
import { User, Credentials } from 'contexts/UserContext/reducer';

import LocalStorageService from './LocalStorageService';

const TOKEN_FIELD_NAME = 'sessionToken';
const TIMEOUT_TIME = 1000;

export interface LoginError {
  message: string;
}

export interface RegistrationUser {
  user: {
    username: string;
    email: string;
    password: string;
  };
}

export const setCurrentUserToken = (currentUser: User) => {
  // TODO: Implement call to authentication API here
  // api.setHeader('Authorization', currentUser.sessionToken);
  LocalStorageService.setValue(TOKEN_FIELD_NAME, currentUser.sessionToken);
};

export const getCurrentUserToken = () => LocalStorageService.getValue(TOKEN_FIELD_NAME);

export const removeCurrentUserToken = () => LocalStorageService.removeValue(TOKEN_FIELD_NAME);

export const login = (credentials: Credentials): Promise<ApiResponse<User, LoginError>> =>
  // TODO: Implement call to authentication API here
  // api.post('/login', credentials);
  new Promise((resolve) => {
    setTimeout(() => {
      resolve({
        ok: true,
        data: { sessionToken: credentials.password === 'asd1' ? 'invalid' : 'token', id: 1234 },
        problem: null,
        originalError: null
      });
    }, TIMEOUT_TIME);
  });

export const signup = (user: RegistrationUser) => api.post('/users', user);

export const logout = (): Promise<ApiResponse<User, LoginError>> =>
  new Promise((resolve) => {
    setTimeout(() => {
      resolve({
        ok: true,
        data: { sessionToken: '', id: 1234 },
        problem: null,
        originalError: null
      });
    }, TIMEOUT_TIME);
  });
