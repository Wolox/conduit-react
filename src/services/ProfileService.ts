import { ApiResponse } from 'apisauce';

import api from 'config/api';
import { PayloadProfile, ResponseProfile } from 'types/profile';

export interface PayloadToggleFollow {
  username: string;
  follow: boolean;
}

const MAIN_PATH = '/profiles';

export const getProfileByUsername = ({ username }: PayloadProfile): Promise<ApiResponse<ResponseProfile>> =>
  api.get(`${MAIN_PATH}/${username}`);

export const toggleFollow = ({
  username,
  follow
}: PayloadToggleFollow): Promise<ApiResponse<ResponseProfile>> =>
  follow ? api.delete(`${MAIN_PATH}/${username}/follow`) : api.post(`${MAIN_PATH}/${username}/follow`);
