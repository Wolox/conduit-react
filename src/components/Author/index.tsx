import { faHeart } from '@fortawesome/free-solid-svg-icons';
import { FontAwesomeIcon } from '@fortawesome/react-fontawesome';
import { memo, useCallback } from 'react';
import { useMutation } from 'react-query';
import cn from 'classnames';
import { generatePath, Link } from 'react-router-dom';

import { SIZE_ICONS_XS } from 'constants/icons';
import { addRemoveFavorites } from 'services/ArticleService';
import PATHS from 'components/Routes/paths';
import { useSelector as useSelectorUser } from 'contexts/UserContext';

import styles from './styles.module.scss';

interface Props {
  image: string;
  username: string;
  date: string;
  favorites: number;
  isFavorited: boolean;
  slug: string;
  following: boolean;
}

function Author({ image, username, date, favorites, isFavorited, slug, following }: Props) {
  const { user } = useSelectorUser((state) => state);
  const { mutate, data } = useMutation(addRemoveFavorites);

  const handleClickFavorited = useCallback(
    (favorite: boolean) => {
      if (user) {
        mutate({ slug, isFavorite: favorite });
      }
    },
    [mutate, slug, user]
  );

  const favorited = data?.data?.article.favorited ?? isFavorited;
  const count = data?.data?.article.favoritesCount ?? favorites;
  const pathToProfile = generatePath(PATHS.profile, { username });
  return (
    <div className={styles.container}>
      <Link to={pathToProfile}>
        <div className={styles.containerAuthor}>
          <img src={image} className={styles.image} />
          <div>
            <span className={styles.username}>{username}</span>
            <span>{date}</span>
            <span data-testId="isUserFollow">{following && '*** You follow this user'}</span>
          </div>
        </div>
      </Link>
      <div className={cn('column center middle', styles.containerButton)}>
        <button
          type="button"
          className={cn(styles.like, { [styles.active]: favorited })}
          onClick={() => handleClickFavorited(favorited)}
          disabled={!user}
        >
          <FontAwesomeIcon icon={faHeart} size={SIZE_ICONS_XS} className={styles.icon} />
          {count}
        </button>
      </div>
    </div>
  );
}

export default memo(Author);
