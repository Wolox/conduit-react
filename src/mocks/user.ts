import { AuthResponse } from 'services/AuthService';
import { UserProfile } from 'types/User';

export const userProfileMock: UserProfile = {
  id: 1,
  username: 'user',
  email: 'user@email.com',
  token: 'token',
  image: 'Man1'
};

export const MOCK_USER: AuthResponse = {
  user: userProfileMock
};
