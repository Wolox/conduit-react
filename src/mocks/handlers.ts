/* eslint-disable no-magic-numbers */
import { rest } from 'msw';

import api, { STATUS_CODES } from 'config/api';

import { MOCKED_ARTICLE_RESPONSE } from './article';
import { userProfileMock } from './user';

const handlers = [
  rest.post('/users', (_req, res, ctx) => res(ctx.status(STATUS_CODES.ok), ctx.json(userProfileMock))),
  rest.post('/users/login', (_req, res, ctx) => res(ctx.status(STATUS_CODES.ok), ctx.json(userProfileMock))),
  rest.post('/articles', (_req, res, ctx) =>
    res(ctx.status(STATUS_CODES.ok), ctx.json(MOCKED_ARTICLE_RESPONSE))
  )
];

export { handlers, rest, api };
