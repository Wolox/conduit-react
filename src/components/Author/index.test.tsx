import { ReactNode } from 'react';
import { render, screen, waitFor } from '@testing-library/react';
import { QueryClient, QueryClientProvider } from 'react-query';
import { Router } from 'react-router';
import { createMemoryHistory } from 'history';
import userEvent from '@testing-library/user-event';

import { server } from 'mocks/server';
import { addRemoveFavorites } from 'services/ArticleService';

import Author from '.';

const MOCK = {
  image: 'conduit.png',
  username: 'conduitWolox',
  date: '07/07/7777',
  favorites: 10,
  isFavorited: true,
  slug: 'titleArticle'
};

function Wrapped({ children }: { children: ReactNode }) {
  const queryClient = new QueryClient();
  const history = createMemoryHistory();
  return (
    <QueryClientProvider client={queryClient}>
      <Router history={history}>{children}</Router>
    </QueryClientProvider>
  );
}

jest.mock('services/ArticleService');

beforeAll(() => server.listen());
afterEach(() => server.resetHandlers());
afterAll(() => server.close());

describe('Component Author', () => {
  test('Renders component', () => {
    const { asFragment } = render(
      <Wrapped>
        <Author {...MOCK} />
      </Wrapped>
    );
    expect(asFragment()).toMatchSnapshot();
  });

  test('The function for add or delete like is called', async () => {
    render(
      <Wrapped>
        <Author {...MOCK} isFavorited={false} />
      </Wrapped>
    );

    userEvent.click(screen.getByRole('button'));
    await waitFor(() =>
      expect(addRemoveFavorites).toHaveBeenCalledWith({ slug: MOCK.slug, isFavorite: false })
    );
  });

  test('The button is active when the isFavorite prop is true', () => {
    render(
      <Wrapped>
        <Author {...MOCK} />
      </Wrapped>
    );
    expect(screen.getByRole('button')).toHaveClass('active');
  });

  test('The button is not active when the isFavorite prop is false', () => {
    render(
      <Wrapped>
        <Author {...MOCK} isFavorited={false} />
      </Wrapped>
    );
    expect(screen.getByRole('button')).not.toHaveClass('active');
  });

  test('Display the total likes', () => {
    render(
      <Wrapped>
        <Author {...MOCK} />
      </Wrapped>
    );
    expect(screen.getByRole('button')).toHaveTextContent('10');
  });
});
