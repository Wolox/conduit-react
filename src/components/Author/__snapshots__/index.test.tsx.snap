// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Component Author Renders component 1`] = `
<DocumentFragment>
  <div
    class="container"
  >
    <div
      class="containerAuthor"
    >
      <img
        class="image"
        src="conduit.png"
      />
      <div>
        <span
          class="username"
        >
          conduitWolox
        </span>
        <span>
          07/07/7777
        </span>
      </div>
    </div>
    <div
      class="column center middle containerButton"
    >
      <button
        class="like active"
        type="button"
      >
        <svg
          aria-hidden="true"
          class="svg-inline--fa fa-heart fa-w-16 fa-xs icon"
          data-icon="heart"
          data-prefix="fas"
          focusable="false"
          role="img"
          viewBox="0 0 512 512"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            d="M462.3 62.6C407.5 15.9 326 24.3 275.7 76.2L256 96.5l-19.7-20.3C186.1 24.3 104.5 15.9 49.7 62.6c-62.8 53.6-66.1 149.8-9.9 207.9l193.5 199.8c12.5 12.9 32.8 12.9 45.3 0l193.5-199.8c56.3-58.1 53-154.3-9.8-207.9z"
            fill="currentColor"
          />
        </svg>
        10
      </button>
    </div>
  </div>
</DocumentFragment>
`;

exports[`Component Author renders component 1`] = `
<DocumentFragment>
  <div
    class="container"
  >
    <div
      class="containerAuthor"
    >
      <img
        class="image"
        src="conduit.png"
      />
      <div>
        <span
          class="username"
        >
          conduitWolox
        </span>
        <span>
          07/07/7777
        </span>
      </div>
    </div>
    <div
      class="column center middle containerButton"
    >
      <button
        class="like active"
        type="button"
      >
        <svg
          aria-hidden="true"
          class="svg-inline--fa fa-heart fa-w-16 fa-xs icon"
          data-icon="heart"
          data-prefix="fas"
          focusable="false"
          role="img"
          viewBox="0 0 512 512"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            d="M462.3 62.6C407.5 15.9 326 24.3 275.7 76.2L256 96.5l-19.7-20.3C186.1 24.3 104.5 15.9 49.7 62.6c-62.8 53.6-66.1 149.8-9.9 207.9l193.5 199.8c12.5 12.9 32.8 12.9 45.3 0l193.5-199.8c56.3-58.1 53-154.3-9.8-207.9z"
            fill="currentColor"
          />
        </svg>
        10
      </button>
    </div>
  </div>
</DocumentFragment>
`;
